### PR TITLE
Adding new dsc extension status cmdlet

### PR DESCRIPTION
**Pull request by vors**
Changes for the new cmdlet - Get-AzureVMDSCExtensionStatus that gets the DSC Extension status for all the vm's deployed in a cloud service or for a particular vm in the service.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@vors__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (sevoroby). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, or work for Microsoft Open Technologies, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+                        Service = vmRoleContext.ServiceName;
+                        VmName = vmRoleContext.Name;
+                        CurrentDeploymentNewSM = this.ComputeClient.Deployments.GetBySlot(vmRoleContext.ServiceName, NSM.DeploymentSlot.Production);
+                    }
+
+                    GetDeploymentOperationNewSM = GetOperationNewSM(CurrentDeploymentNewSM.RequestId);
+                    WriteVerboseWithTimestamp(Resources.GetDeploymentCompletedOperation);
+                }
+                catch (CloudException ex)
```
Comment (markcowl) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs]: No longer CloudException

**Comment by markcowl** [comment]
Need to update the new cmdlet for the changes in the latest common library (for example CloudException -> HttpOperationException)

**Comment by azurecla** [comment]
__@vors__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

**Comment by stankovski** [comment]
@vors - we are planning to discontinue mstest runner in the near future. Please update your tests to use XUnit

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    /// in a cloud service. When a configuration is applied this cmdlet produces output consistent with Start-DscConfiguration. 
+    /// 
+    ///  Not: To get detailed output -Verbose flag need to be specified
+    /// 
+    /// Example Usage:
+    /// Get-AzureVMDscExtensionStatus -ServiceName service
+    /// Get-AzureVMDscExtensionStatus -ServiceName service -Name VM-name
+    /// Get-AzureVMDscExtensionStatus -VM vm
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, VirtualMachineDscStatusCmdletNoun, DefaultParameterSetName = GetStatusByServiceAndVmNameParamSet), OutputType(typeof(VirtualMachineDscExtensionStatusContext))]
```
Comment (stankovski) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs]: Wrap long lines for readability

File: src/ServiceManagement/Compute/Commands.ServiceManagement.Extensions.Test/Commands.ServiceManagement.Extensions.Test.csproj [hunk 3]
```diff
@@ -70,6 +73,14 @@
     <Reference Include="Microsoft.Threading.Tasks.Extensions.Desktop">
       <HintPath>..\..\..\packages\Microsoft.Bcl.Async.1.0.168\lib\net40\Microsoft.Threading.Tasks.Extensions.Desktop.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.WindowsAzure.Common, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (matt-gibbs) [inline src/ServiceManagement/Compute/Commands.ServiceManagement.Extensions.Test/Commands.ServiceManagement.Extensions.Test.csproj]: why reference the old version?

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                }
+            });
+        }
+
+        /// <summary>
+        /// Retrieves dsc extension status for all virtual machine's in a cloud service or a given virtual machine from the deployment object
+        /// </summary>
+        /// <param name="deployment">Deployment that exists in the service</param>
+        /// <returns></returns>
+        internal List<VirtualMachineDscExtensionStatusContext> GetVirtualMachineDscStatusContextList(NSM.DeploymentGetResponse deployment)
```
Comment (stankovski) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs]: Please move all business logic to a convenience client (see https://github.com/Azure/azure-powershell/blob/master/src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/ResourceClient.cs)

File: src/ServiceManagement/Compute/Commands.ServiceManagement.Extensions.Test/Commands.ServiceManagement.Extensions.Test.csproj [hunk 5]
```diff
@@ -107,19 +118,29 @@
...
+      <Project>{4900EC4E-8DEB-4412-9108-0BC52F81D457}</Project>
+      <Name>Commands.Utilities</Name>
+    </ProjectReference>
     <ProjectReference Include="..\Commands.ServiceManagement\Commands.ServiceManagement.csproj">
       <Project>{e1ca72ba-8374-45f6-904d-fd34ecdf5b6f}</Project>
       <Name>Commands.ServiceManagement</Name>
     </ProjectReference>
   </ItemGroup>
   <ItemGroup>
+    <None Include="MSSharedLibKey.snk" />
```
Comment (matt-gibbs) [inline src/ServiceManagement/Compute/Commands.ServiceManagement.Extensions.Test/Commands.ServiceManagement.Extensions.Test.csproj]: Please check with Yugang on delay signing and directly referencing the snk file.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtension.cs [hunk 6]
```diff
@@ -12,16 +12,16 @@
 // limitations under the License.
 // ----------------------------------------------------------------------------------
 
+using Microsoft.WindowsAzure.Commands.ServiceManagement.Helpers;
```
Comment (matt-gibbs) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtension.cs]: coding convention is to put the System references first.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        /// <param name="roleInstance">Details of a specific role instance</param>
+        /// <returns></returns>
+        internal VirtualMachineDscExtensionStatusContext CreateDscStatusContext(NSM.Role vmRole, NSM.RoleInstance roleInstance) 
+        {
+            var message = string.Empty;
+            NSM.ResourceExtensionConfigurationStatus extensionSettingStatus = null;
+            
+            if (roleInstance != null && roleInstance.ResourceExtensionStatusList != null)
+            {
+                foreach (var resourceExtensionStatus in roleInstance.ResourceExtensionStatusList.Where(resourceExtensionStatus => resourceExtensionStatus.HandlerName.Equals(VirtualMachineDscExtensionCmdletBase.ExtensionPublishedNamespace + "." + VirtualMachineDscExtensionCmdletBase.ExtensionPublishedName, StringComparison.InvariantCultureIgnoreCase)).Where(resourceExtensionStatus => resourceExtensionStatus.ExtensionSettingStatus != null))
```
Comment (matt-gibbs) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs]: 440+ characters on a line is excessive. Split appropriately.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs (hunk 4 above)
Comment (eshaparmar) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/DSC/GetAzureVMDscExtensionStatus.cs]: Are you referring to creating a separate common class for the business logic for all DSC cmdlets or just this cmdlet? The logic and data manipulation done in this cmdlet is not going to be used by elsewhere. So, I feel that putting it in a different class is not going to be of any use.

File: src/ServiceManagement/Compute/Commands.ServiceManagement.Extensions.Test/Commands.ServiceManagement.Extensions.Test.csproj (hunk 5 above)
Comment (eshaparmar) [inline src/ServiceManagement/Compute/Commands.ServiceManagement.Extensions.Test/Commands.ServiceManagement.Extensions.Test.csproj]: Spoke with Yugang offline - once there is a common place for the .snk file all duplicates will be removed. We thought about putting a duplicate .snk file but decided to go with the pattern that is followed so far ( to be consistent). Ideally, yes there should be a common place for the .snk file.

**Pull request merged**